src/cosim/execution.cpp

```diff
@@ -549,6 +560,22 @@ entity_index_maps inject_system_structure(
             val);
     }
 
+    // For ECCO algorithm, add the powerbonds
+    if (sys.algorithm == "ecco") {
+        const auto powerbonds = sys.get_power_bonds();
+        auto algo = (cosim::algorithm*)exe.get_algorithm().get();
+        auto algorithm = reinterpret_cast<cosim::ecco_algorithm*>(algo); /// ugly downcasting, better alternatives?
```

Comment:
Yes, use [`std::dynamic_pointer_cast()`](https://en.cppreference.com/w/cpp/memory/shared_ptr/pointer_cast). In light of my request to remove `system_structure::algorithm` again, however, I suggest a change to the entire `if` block – something like this:
```cpp
// For ECCO algorithm, add the power bonds
if (auto algorithm = std::dynamic_pointer_cast<ecco_algorithm>(exe.get_algorithm())) {
    const auto powerBonds = sys.get_power_bonds();
    if (powerBonds.empty()) {
        throw /* some exception saying that ECCO won't work without power bonds */;
    }
    // ...
```